jcstress-samples/src/main/java/org/openjdk/jcstress/samples/concurrency/classic/ClassicProblem_02_ProducerConsumerProblem.java

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jcstress.samples.concurrency.classic;
+
+import org.openjdk.jcstress.annotations.*;
+import org.openjdk.jcstress.infra.results.III_Result;
+import org.openjdk.jcstress.infra.results.Z_Result;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE;
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE_INTERESTING;
+
+/*
+    How to run this test:
+        $ java -jar jcstress-samples/target/jcstress.jar -t ClassicProblem_02_ProducerConsumerProblem
+ */
+
+/**
+ * This sample shows you how JCStress can help you to test solutions for the famous producer-consumer problem.
+ * See https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem for more information about the problem and solutions.
+ */
+public abstract class ClassicProblem_02_ProducerConsumerProblem {
+    private final static int BUFFER_SIZE = 2;
+
+    static class SemaphoresBase {
+        protected final Semaphore fillCount = new Semaphore(BUFFER_SIZE);
+        protected final Semaphore emptyCount = new Semaphore(BUFFER_SIZE);
+
+        public SemaphoresBase() {
+            try {
+                fillCount.acquire(BUFFER_SIZE);
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public int produce() {
+            try {
+                // produce item
```

Comment:
What these comments are for? How do they relate to test?

Reply:
They are placeholders for code which was left out because it isn't needed for the test in JCStress. However, users of JCStress must not forget to add code there if they want to implement e.g. a queue with a buffer. The solutions in the sample are only the "bones" of the algorithms. The "muscles" are missing. (Excuse my strange metaphor here)

Reply:
I have noticed just a moment ago that the original solutions in the wikipedia artcle (https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem) also use comments and undeclared functions to show readers which code was left out.